### DISCUSSION TITLE
Plugable optics emulation (SFP, QSFP, ...)

### DISCUSSION THREAD

**Discussion by insekt** [ipspace/netlab#1530] [Features]
Is there a way to emulate optical modules like SFP, SFP+, QSFP, ...? I.e. so that the emulator can work with DDMI parameters (rx power, temp., ...), LOS, Tx_Fault signals. I have this problem for a long time, because in a real network, a networking problem with can be caused by a transceiver, a low signal level, a high temperature. It would be very helpful to simulate some physical level problems.
I do not see a way to integrate this into virtual device images. But maybe there is a way to organize an additional emulation layer related to links? I would appreciate any ideas. I am willing to spend time on analysis and prototyping.

**Comment by jbemmel** [comment]
You can emulate the effects of physical issues such as packet loss, latency variations and combinations thereof, but direct emulation will be difficult as most platforms are closed source. Perhaps in the Linux kernel you could experiment with returning different values for device level parameters like temperature readings - but for typical network equipment that will be difficult

**Comment by Astrotokii** [comment]
In addition, it would be interesting to simulate multimode, singlemode, bidirectional transceivers, etc. for fiber connections. Like maybe emulate FlexOptix universal transceiver so compatibility wouldn't be an issue between hosts?

**Comment by ipspace** [comment]
Emulating link impairment is relatively easy (configure **tc** parameters on an interface or on a Linux bridge, see also https://github.com/ipspace/netlab/issues/1387). Anything beyond that seems to be Mission Impossible.
